### PR TITLE
Parse instanced meshes in each sector, instead of putting all of them into the first sector containing a given file id

### PR DESCRIPTION
**Pull request by Verkhovskaya**

### CODE REVIEW

**Comment by anderscognite** [comment]
I think it downloads more than it should as I get this in the console:

`{step: 3, numSteps: 3, title: "Loading additional details...", progress: 14815, max: 72111}`

and it's very slow :D

File: src/customFileParser/unpackGeometry/InstancedMesh.ts [hunk 1]
```diff
@@ -19,16 +19,16 @@ export default function unpackInstancedMeshes(
   ) {
 
   const data = new PropertyLoader(uncompressedValues);
-  const meshCounts: {[fileId: string]: {[triangleOffset: string]: { count: number, triangleCount: number }}} = {};
-  const fileIdToSector: {[fileId: string]: Sector} = {};
 
-  // Count meshes per file Id and triangle offset
-  for (const sector of rootSector.traverseSectors()) {
+  for (const sector of rootSector!.traverseSectors()) {
```
Comment (anderscognite) [inline src/customFileParser/unpackGeometry/InstancedMesh.ts]: Why did anything change here? rootSector is not an optional argument into the function.

File: src/customFileParser/unpackGeometry/InstancedMesh.ts [hunk 2]
```diff
@@ -19,16 +19,16 @@ export default function unpackInstancedMeshes(
...
+    sector.instancedMeshGroup = new InstancedMeshGroup();
+    const meshCounts: {[fileId: string]: {[triangleOffset: string]: { count: number, triangleCount: number }}} = {};
+
+    // Count meshes per file Id and triangle offset
```
Comment (anderscognite) [inline src/customFileParser/unpackGeometry/InstancedMesh.ts]:
```suggestion
    // Count meshes per file id and triangle offset
```

File: src/customFileParser/unpackGeometry/InstancedMesh.ts [hunk 3]
```diff
@@ -37,22 +37,19 @@ export default function unpackInstancedMeshes(
...
-  Object.keys(meshCounts).forEach(fileId => {
-    collections[fileId] = collections[fileId] ? collections[fileId] : {};
-    Object.keys(meshCounts[fileId]).forEach(triangleOffset => {
-      const { count, triangleCount } = meshCounts[fileId][triangleOffset];
-      collections[fileId][triangleOffset] = new InstancedMeshCollection(
-        parseInt(triangleOffset, 10), triangleCount, count);
+    // Create mesh collections for each file Id and triangle offset
+    const collections: {[fileId: string]: {[triangleOffset: string]: InstancedMeshCollection}} = {};
+    Object.keys(meshCounts).forEach(fileId => {
+      collections[fileId] = collections[fileId] ? collections[fileId] : {};
```
Comment (anderscognite) [inline src/customFileParser/unpackGeometry/InstancedMesh.ts]: I like being explicit here with a check with `undefined` or `null`, and not any truthy value.

File: src/customFileParser/unpackGeometry/InstancedMesh.ts [hunk 4]
```diff
@@ -63,20 +60,19 @@ export default function unpackInstancedMeshes(
...
-  }
 
-  // Add collections to sector group
-  Object.keys(collections).forEach(fileId => {
-    const instancedMesh = new InstancedMesh(parseInt(fileId, 10));
-    Object.keys(collections[fileId]).forEach(triangleOffset => {
-      instancedMesh.addCollection(collections[fileId][triangleOffset]);
+    // Add collections to sector group
+    Object.keys(collections).forEach(fileId => {
+      const instancedMesh = new InstancedMesh(parseInt(fileId, 10));
```
Comment (anderscognite) [inline src/customFileParser/unpackGeometry/InstancedMesh.ts]: I'm not too happy about all the parseInt's everywhere. Shouldn't we have converted these values to numbers once we parsed them instead of having to parseInt a lot of places?

**Review by anderscognite — CHANGES_REQUESTED** [review]
Sweet! Some questions of course :D

**Comment by Verkhovskaya** [comment]
Re: 'Downloading' way too many files, it's because you need a corresponding change to 3d-viewer that'll merge all the newly created instancedMesh meshes. That PR is in progress (I wasn't going to merge this code into master until it was also reviewed/confirmed)

File: src/customFileParser/unpackGeometry/InstancedMesh.ts (hunk 1 above)
Comment (Verkhovskaya) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: A case of bad copy/paste. Fixed.

File: src/customFileParser/unpackGeometry/InstancedMesh.ts (hunk 3 above)
Comment (Verkhovskaya) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: Fixed

File: src/customFileParser/unpackGeometry/InstancedMesh.ts (hunk 4 above)
Comment (Verkhovskaya) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: I could do something like forEach(triangleOffsetStr => {
const triangleOffset = parseInt(triangleOffsetStr, 10);
func(triangleOffset)
}

but it seems redundant, since I'm still only using triangleOffset once. The parseInts are necessary because JavaScript object keys are always converted to strings.

Comment (anderscognite) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: More generally, a type that is `number` should be `number` after a parser has done its job, shouldn't it? But I was wrong, it is not everywhere, only two places.

Comment (Verkhovskaya) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: Not quite sure what you mean. If I had something like
const myObject = {1: 'hello'}
myObject.keys() would return ['1'], not [1]. It's a JavaScript thing.

Comment (anderscognite) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: Omg sorry, my bad. You've told me this before :D

Comment (Verkhovskaya) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: Fricken JavaScript. Sounds like I'm good to keep this as-is?

Comment (anderscognite) [reply src/customFileParser/unpackGeometry/InstancedMesh.ts]: Yes!

**Comment by anderscognite** [comment]
Can you merge in master before review as I need to review after that anyway :D

File: src/geometry/GeometryGroupData.ts [hunk 5]
```diff
@@ -1,12 +1,12 @@
 // Copyright 2019 Cognite AS
 
 import { float64Properties, float32Properties, vector3Properties, matrix4Properties,
-  primitiveName, primitiveProperties, propertyName, colorProperties, vector4Properties,
-  primitiveAttributeProperties }
+  renderedPrimitiveNameType, primitiveProperties, renderedPropertyNameType, colorProperties, vector4Properties,
```
Comment (nsunderland-cognite) [inline src/geometry/GeometryGroupData.ts]: What's the idea behind the "rendered" naming?

File: src/geometry/MergedMeshGroup.ts [hunk 6]
```diff
@@ -9,6 +9,31 @@ interface IndexMap { [s: number]: boolean; }
 const globalBox = new THREE.Box3();
 
 export class MergedMeshMappings {
+  public static concat(mappings: MergedMeshMappings[]) {
+    let capacity = 0; mappings.forEach(mapping => { capacity += mapping.count; });
```
Comment (nsunderland-cognite) [inline src/geometry/MergedMeshGroup.ts]: `const capacity = mappings.reduce((sum, mapping) => sum + mapping.count, 0);`

File: src/geometry/MergedMeshGroup.ts [hunk 7]
```diff
@@ -199,11 +240,11 @@ export class MergedMeshGroup extends GeometryGroup {
       const mergedMesh = this.meshes[meshIndex];
 
       if (geometry == null) {
-        if (mergedMesh.geometry == null) {
+        if (this.geometry == null) {
```
Comment (nsunderland-cognite) [inline src/geometry/MergedMeshGroup.ts:243]: Why was this changed?

**Comment by Verkhovskaya** [comment]
Oh, this is fun. Basically, we used to have a MergedMeshGroup with lots of meshes, and each mesh had a separate object. Now, we have 1 mesh per object, so the .geometry (aka the mesh) is attached to MergedMeshGroup now, instead of to each MergedMesh

File: src/optimizations/optimizeMeshes.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+  for (const sector of rootSector.traverseSectorsBreadthFirst()) {
+    sector.instancedMeshGroup.meshes.forEach(instancedMesh => {
+      const mergedMesh = new MergedMesh(
+        countMappingsToMerge(instancedMesh.collections),
+        instancedMesh.fileId,
+        true,
+      );
+
+      for (let index = instancedMesh.collections.length - 1; index >= 0; index--) {
```
Comment (nsunderland-cognite) [inline src/optimizations/optimizeMeshes.ts]: Aaaaargh what kind of language doesn't have a backwards iterator

File: src/parsers/i3d/PropertyLoader.ts [hunk 9]
```diff
@@ -91,6 +92,8 @@ export default class PropertyLoader {
     'triangleCount':  (geometry: CGD) => { this.triangleCount                = geometry.indices.nextValue() ; },
     'thickness':      (geometry: CGD) => { this.thickness = this.values.radius![geometry.indices.nextValue()]; },
     'fileId':         (geometry: CGD) => { this.fileId    = this.values.fileId![geometry.indices.nextValue()]; },
+    'diagonalSize':   (geometry: CGD) => {
```
Comment (nsunderland-cognite) [inline src/parsers/i3d/PropertyLoader.ts]: What on Earth is a CGD? Why do we have a type named like this?

File: src/parsers/i3d/loadCompressedGeometryData.ts [hunk 10]
```diff
@@ -9,7 +9,12 @@ export default function loadGeometryIndices(file: CustomFileReader, sectorEndLoc
 
   const geometryIndices: CompressedGeometryData[] = [];
   while (file.location < sectorEndLocation) {
-    const type = IdToFileGeometryName[file.readUint8()];
+    const typeId = file.readUint8();
+    const type = IdToFileGeometryName[typeId];
+    if (type === undefined) {
+      // This will happen if file type 4 is used
```
Comment (nsunderland-cognite) [inline src/parsers/i3d/loadCompressedGeometryData.ts:15]: What's the deal with this?

File: src/geometry/MergedMeshGroup.ts (hunk 7 above)
Comment (anderscognite) [reply src/geometry/MergedMeshGroup.ts:243]: I think having the geometry on the MergedMesh makes sense though? At least until all meshes are merged into one in this repository. The data structure here assumes an optimization that doesn't necessarily need to happen.

File: src/geometry/GeometryGroupData.ts (hunk 5 above)
Comment (Verkhovskaya) [reply src/geometry/GeometryGroupData.ts]: So, there are file primitives, and rendered primitives, and they're different. For example, a closedCylinder might be rendered as one cylinder and two circles

File: src/geometry/MergedMeshGroup.ts (hunk 6 above)
Comment (Verkhovskaya) [reply src/geometry/MergedMeshGroup.ts]: Cheers

File: src/geometry/MergedMeshGroup.ts (hunk 7 above)
Comment (Verkhovskaya) [reply src/geometry/MergedMeshGroup.ts:243]: It does necessarily need to happen. When I turn instanced meshes into merged meshes, I end up with a few dozen extra, tiny meshes. So I turn ALL of the meshes in a mergedMeshGroup into one object.

Comment (Verkhovskaya) [reply src/geometry/MergedMeshGroup.ts:243]: Tl;dr: this lets us draw all the merged meshes in a model with 1 draw call per sector

Comment (Verkhovskaya) [reply src/geometry/MergedMeshGroup.ts:243]: Talked to Anders, we're going to move everything into a single mergedMesh and delete the old meshes. Will come in a later PR, since I'm re-writing how we merge meshes anyways.

File: src/parsers/i3d/PropertyLoader.ts (hunk 9 above)
Comment (Verkhovskaya) [reply src/parsers/i3d/PropertyLoader.ts]: CompressedGeometryData, defined at the top of the file. It's a long name and I wanted everything to line up nicely :)

File: src/parsers/i3d/loadCompressedGeometryData.ts (hunk 10 above)
Comment (Verkhovskaya) [reply src/parsers/i3d/loadCompressedGeometryData.ts:15]: You'll notice that I have a bunch of console.warns everywhere. They basically yell at you if the parser parameters are wrong.

Comment (Verkhovskaya) [reply src/parsers/i3d/loadCompressedGeometryData.ts:15]: But long story with file versions. I'll take this out once I make file type 5 the default.

File: src/parsers/i3d/PropertyLoader.ts (hunk 9 above)
Comment (nsunderland-cognite) [reply src/parsers/i3d/PropertyLoader.ts]: I disagree with this. Type names should be explicit and clear. CGD is confusing.

Comment (Verkhovskaya) [reply src/parsers/i3d/PropertyLoader.ts]: Normally I'd agree with you, but this file is very repetitive, so I formatted it accordingly. Notice the //tslint: disable at the top

Comment (Verkhovskaya) [reply src/parsers/i3d/PropertyLoader.ts]: Readability over everything

Comment (anderscognite) [reply src/parsers/i3d/PropertyLoader.ts]: I don't see why this makes it more readable. `CompressedGeometryData` is super easy to read. I agree with @nsunderland-cognite here.

**Comment by anderscognite** [comment]
Travis is not happy. Does `yarn lint` work locally @Verkhovskaya? Could you fix the GCD naming, make sure travis is happy, and I'll review.

**Comment by nsunderland-cognite** [comment]
> Travis is not happy. Does `yarn lint` work locally @Verkhovskaya? Could you fix the GCD naming, make sure travis is happy, and I'll review.

Doesn't this project have a prepush hook for linting?

**Comment by anderscognite** [comment]
Tests are failing:
```
 FAIL  src/__tests__/SphericalSegmentGroup.test.ts
  ● Test suite failed to run
    TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
    src/__tests__/SphericalSegmentGroup.test.ts:60:5 - error TS2554: Expected 7-8 arguments, but got 6.
    60     group.add(nodeId, treeIndex, center, normal, radius, height);
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      src/geometry/SphericalSegmentGroup.ts:55:5
        55     height: number,
               ~~~~~~~~~~~~~~
        An argument for 'height' was not provided.
```

**Comment by Verkhovskaya** [comment]
Fixed the tests :)

File: src/__tests__/BoxGroup.test.ts [hunk 11]
```diff
@@ -12,7 +12,8 @@ function createBoxGroup(): BoxGroup {
   const normal = new THREE.Vector3(4, 5, 6);
   const angle = 1.0;
   const delta = new THREE.Vector3(1, 1, 1);
-  group.add(nodeId, treeIndex, center, normal, angle, delta);
+  const diagonalSize = Math.sqrt(delta.x ** 2 + delta.y ** 2 + delta.z ** 2);
```
Comment (anderscognite) [inline src/__tests__/BoxGroup.test.ts]: I think we talked about naming this just `size` or something? As diagonal size isn't correct for all objects (it is here though), or is it?

File: src/__tests__/CircleGroup.test.ts [hunk 12]
```diff
@@ -12,8 +12,9 @@ function createCircleGroup(): CircleGroup {
   const center = new THREE.Vector3(1, 2, 3);
   const normal = new THREE.Vector3(4, 5, 6);
   const radius = 10.0;
+  const diagonalSize = radius;
```
Comment (anderscognite) [inline src/__tests__/CircleGroup.test.ts]: Here for instance, the correct value should be 2 * radius to reflect the size of the object on the screen (which is what we assume later on).

Comment (anderscognite) [inline src/__tests__/CircleGroup.test.ts]:
```suggestion
  const diagonalSize = 2 * radius;
```

File: src/__tests__/CircleGroup.test.ts [hunk 13]
```diff
@@ -41,8 +42,9 @@ describe('CircleGroup', () => {
     const center = new THREE.Vector3(1, 2, 3);
     const normal = new THREE.Vector3(4, 5, 6);
     const radius = 10.0;
+    const diagonalSize = radius;
```
Comment (anderscognite) [inline src/__tests__/CircleGroup.test.ts]:
```suggestion
    const diagonalSize = 2 * radius;
```

File: src/__tests__/fixtures/Optimized/uploaded_files.txt [hunk 14]
```diff
@@ -0,0 +1,29 @@
+642282308	processed.opt.ciff
```
Comment (anderscognite) [inline src/__tests__/fixtures/Optimized/uploaded_files.txt]: It is unclear to me what `Optimized` means. I think the directory should give a hint of what it is.

File: src/__tests__/i3dParser/integrationTest.ts [hunk 15]
```diff
@@ -83,11 +83,12 @@ describe('customFileIntegrationTest', () => {
     }
     const compressedPrimitiveData = fileReader.readCompressedGeometryData(fileBuffer.byteLength).primitives;
 
+    // TODO(verkhovskaya): Get a better file for testing this
     // Check that the sector has geometries. If it doesn't, run this test on a different file.
-    expect(compressedPrimitiveData!.length).toBeGreaterThan(0);
+    // expect(compressedPrimitiveData!.length).toBeGreaterThan(0);
```
Comment (anderscognite) [inline src/__tests__/i3dParser/integrationTest.ts]: We should either take out the test, or leave it in.

File: src/__tests__/i3dParser/integrationTest.ts [hunk 16]
```diff
@@ -3,14 +3,14 @@
 import * as THREE from 'three';
 import CustomFileReader from '../../parsers/i3d/CustomFileReader';
 import { parseFullCustomFile } from '../../parsers/i3d/main';
-import { filePrimitiveNames, filePropertyArrayNames, filePropertyArrayNameType }
+import { filePrimitiveNames, filePropertyArrayNames, filePropertyArrayNameType, filePrimitiveNameType }
```
Comment (anderscognite) [inline src/__tests__/i3dParser/integrationTest.ts]: Types should be upper camel case.

File: src/geometry/CircleGroup.ts [hunk 17]
```diff
@@ -47,13 +47,15 @@ export default class CircleGroup extends PrimitiveGroup {
   add(
     nodeId: number,
     treeIndex: number,
+    diagonalSize: number,
```
Comment (anderscognite) [inline src/geometry/CircleGroup.ts]: Rename to size.

File: src/geometry/GeometryGroupData.ts (hunk 5 above)
Comment (anderscognite) [inline src/geometry/GeometryGroupData.ts]: Types should be upper camel case.

File: src/geometry/GeometryGroupData.ts [hunk 18]
```diff
@@ -97,15 +97,19 @@ export default class GeometryGroupData {
 
   add(properties: any) {
     Object.keys(properties).forEach(property => {
-      const name = property as propertyName;
+      if (this.arrays[property] === undefined) {
+        throw Error('Property ' + (property as string) + ' is not present on groups with type ' +
+          (this.type as string) + '. See primitiveProperties in GeometryGroupDataParameters.ts');
```
Comment (anderscognite) [inline src/geometry/GeometryGroupData.ts:102]: Why do you need to do `as string`? Isn't it a `string`?

File: src/geometry/GeometryGroupData.ts [hunk 19]
```diff
@@ -97,15 +97,19 @@ export default class GeometryGroupData {
...
+      }
+      const name = property as renderedPropertyNameType;
```
Comment (anderscognite) [inline src/geometry/GeometryGroupData.ts]: Types should be in upper camel case.

File: src/geometry/GeometryGroupDataParameters.ts [hunk 20]
```diff
@@ -44,45 +46,13 @@ export const primitiveAttributeProperties: { [name in primitiveName]: propertyNa
...
-triangleCount?: any;
-vertex1?: any;
-vertex2?: any;
-vertex3?: any;
-vertex4?: any;
-}
-
-export const float64Properties: propertyName[] = ['nodeId'];
-export const float32Properties: propertyName[] =
+export const float64Properties: renderedPropertyNameType[] = ['nodeId'];
```
Comment (anderscognite) [inline src/geometry/GeometryGroupDataParameters.ts]: Types should be upper camel case.

File: src/geometry/InstancedMeshGroup.ts [hunk 21]
```diff
@@ -159,6 +178,7 @@ export class InstancedMeshCollection {
   triangleOffset: number;
   triangleCount: number;
   geometry?: THREE.InstancedBufferGeometry;
+  geometrySizes?: Float32Array;
```
Comment (anderscognite) [inline src/geometry/InstancedMeshGroup.ts]: Not used anywhere? Why do we have an array when we have mappings?

File: src/geometry/MergedMeshGroup.ts [hunk 22]
```diff
@@ -9,6 +9,31 @@ interface IndexMap { [s: number]: boolean; }
 const globalBox = new THREE.Box3();
 
 export class MergedMeshMappings {
+  public static concat(mappings: MergedMeshMappings[]) {
```
Comment (anderscognite) [inline src/geometry/MergedMeshGroup.ts:12]: Should we abbreviate this or have `concatenate` in full? @dragly @nsunderland-cognite

File: src/geometry/MergedMeshGroup.ts [hunk 23]
```diff
@@ -9,6 +9,31 @@ interface IndexMap { [s: number]: boolean; }
...
+    const capacity = mappings.reduce((sum, mapping) => sum + mapping.count, 0);
+    const newMapping = new MergedMeshMappings(capacity);
+    newMapping.count = capacity;
+    let offset = 0;
```
Comment (anderscognite) [inline src/geometry/MergedMeshGroup.ts]: Since we have another offset, maybe specify what kind of offset this is.

File: src/geometry/MergedMeshGroup.ts [hunk 24]
```diff
@@ -20,6 +45,8 @@ export class MergedMeshMappings {
   public transform2: Float32Array[];
   public transform3: Float32Array[];
```
Comment (anderscognite) [inline src/geometry/MergedMeshGroup.ts:47]:
```suggestion
```

File: src/geometry/MergedMeshGroup.ts [hunk 25]
```diff
@@ -30,13 +57,16 @@ export class MergedMeshMappings {
     this.transform1 = []; this.transform1.length = capacity;
     this.transform2 = []; this.transform2.length = capacity;
     this.transform3 = []; this.transform3.length = capacity;
+
```
Comment (anderscognite) [inline src/geometry/MergedMeshGroup.ts:60]:
```suggestion
```

File: src/geometry/MergedMeshGroup.ts [hunk 26]
```diff
@@ -117,11 +152,15 @@ export class MergedMeshMappings {
...
+  public setDiagonalSize(value: number, index: number) {
+    this.diagonalSize[index] = value;
+  }
+
   public setTransform(source: THREE.Matrix4, index: number) {
     const columns = [0, 1, 2, 3].map(columnIndex => {
       // @ts-ignore
       const transform = this[`transform${columnIndex}`];
-      return transform[this.count] = new Float32Array(3);
+      return transform[index] = new Float32Array(3);
```
Comment (anderscognite) [inline src/geometry/MergedMeshGroup.ts:163]: Nice!

File: src/optimizations/optimizeMeshes.ts [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+  collections.forEach(collection => {
+    const triangleCount = collection.mappings.count * collection.triangleCount;
+    if (triangleCount < TRIANGLE_COUNT_LIMIT) {
+      numMappings += collection.mappings.count;
+    }
+  });
+  return numMappings;
+}
+
+export default function optimizeMeshes(
```
Comment (anderscognite) [inline src/optimizations/optimizeMeshes.ts]: I still think this is a bad function name unless it does a series of optimizations that do have good names. `mergeInstancedMeshes` is telling you more about what's happening.

File: src/parsers/i3d/CustomFileReader.ts [hunk 28]
```diff
@@ -5,7 +5,7 @@ import loadUncompressedValues from './loadUncompressedValues';
 import loadCompressedGeometryData from './loadCompressedGeometryData';
 import FibonacciDecoder from '../FibonacciDecoder';
 import { NodeIdReader, CompressedGeometryData, SectorCompressedData } from './sharedFileParserTypes';
-import { filePrimitiveNames, BYTES_PER_NODE_ID, geometryNameType } from './parserParameters';
+import { filePrimitiveNames, BYTES_PER_NODE_ID, fileGeometryNameType, filePrimitiveNameType } from './parserParameters';
```
Comment (anderscognite) [inline src/parsers/i3d/CustomFileReader.ts]: Types should be in upper camel case.

File: src/parsers/i3d/loadCompressedGeometryData.ts (hunk 10 above)
Comment (anderscognite) [reply src/parsers/i3d/loadCompressedGeometryData.ts:15]: Could you make a clear TODO here (so it's searchable)

File: src/parsers/i3d/unpackGeometry/InstancedMesh.ts [hunk 29]
```diff
@@ -1,13 +1,13 @@
 // Copyright 2019 Cognite AS
 
 import * as THREE from 'three';
-import { InstancedMesh, InstancedMeshCollection } from '../../../geometry/InstancedMeshGroup';
-import { PerSectorCompressedData, UncompressedValues } from '../sharedFileParserTypes';
-import PropertyLoader from '../PropertyLoader';
+import { InstancedMesh, InstancedMeshCollection, InstancedMeshGroup } from '../../../geometry/InstancedMeshGroup';
+import { PerSectorCompressedData, UncompressedValues } from './../sharedFileParserTypes';
+import PropertyLoader from './../PropertyLoader';
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/InstancedMesh.ts]: Why do you have `./../`?

Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/InstancedMesh.ts]:
```suggestion
import PropertyLoader from '../PropertyLoader';
```

File: src/parsers/i3d/unpackGeometry/InstancedMesh.ts [hunk 30]
```diff
@@ -1,13 +1,13 @@
 // Copyright 2019 Cognite AS
 
 import * as THREE from 'three';
-import { InstancedMesh, InstancedMeshCollection } from '../../../geometry/InstancedMeshGroup';
-import { PerSectorCompressedData, UncompressedValues } from '../sharedFileParserTypes';
-import PropertyLoader from '../PropertyLoader';
+import { InstancedMesh, InstancedMeshCollection, InstancedMeshGroup } from '../../../geometry/InstancedMeshGroup';
+import { PerSectorCompressedData, UncompressedValues } from './../sharedFileParserTypes';
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/InstancedMesh.ts]:
```suggestion
import { PerSectorCompressedData, UncompressedValues } from '../sharedFileParserTypes';
```

File: src/parsers/i3d/unpackGeometry/InstancedMesh.ts [hunk 31]
```diff
@@ -1,13 +1,13 @@
...
+import PropertyLoader from './../PropertyLoader';
 import { xAxis, yAxis, zAxis } from '../../../constants';
 import SceneStats from '../../../SceneStats';
 import Sector from '../../../Sector';
-import { DataMaps } from '../../parseUtils';
+import { DataMaps } from '../../../parsers/parseUtils';
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/InstancedMesh.ts]: Hmm?

Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/InstancedMesh.ts]:
```suggestion
import { DataMaps } from '../../parseUtils';
```

File: src/parsers/i3d/PropertyLoader.ts (hunk 9 above)
Comment (Verkhovskaya) [reply src/parsers/i3d/PropertyLoader.ts]: Talked to Nick/Anders, moved code around

File: src/parsers/i3d/unpackGeometry/main.ts [hunk 32]
```diff
@@ -3,8 +3,9 @@
 import { PerSectorCompressedData, UncompressedValues, CompressedGeometryData }
   from './../sharedFileParserTypes';
 import PropertyLoader from './../PropertyLoader';
-import { renderedPrimitiveToAddFunction, renderedPrimitivesPerFilePrimitive, renderedPrimitiveToGroup,
-  renderedPrimitiveNames} from '../parserParameters';
+import { renderedPrimitiveToAddFunction, renderedPrimitivesPerFilePrimitive, renderedPrimitiveToGroup }
+  from '../parserParameters';
+import { renderedPrimitiveNames } from '../../../geometry/GeometryGroupDataParameters';
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/main.ts]: Is this a type too? Upper camel case!

**Review by anderscognite — CHANGES_REQUESTED** [review]
Good stuff, but a couple of things!

File: src/geometry/InstancedMeshGroup.ts [hunk 33]
```diff
@@ -20,6 +20,8 @@ export class InstancedMeshMappings {
   public transform2: Float32Array;
   public transform3: Float32Array;
```
Comment (anderscognite) [inline src/geometry/InstancedMeshGroup.ts]:
```suggestion
```

File: src/geometry/InstancedMeshGroup.ts [hunk 34]
```diff
@@ -30,6 +32,8 @@ export class InstancedMeshMappings {
     this.transform1 = new Float32Array(3 * this.capacity);
     this.transform2 = new Float32Array(3 * this.capacity);
     this.transform3 = new Float32Array(3 * this.capacity);
+
```
Comment (anderscognite) [inline src/geometry/InstancedMeshGroup.ts]:
```suggestion
```

File: src/geometry/MergedMeshGroup.ts (hunk 22 above)
Comment (anderscognite) [reply src/geometry/MergedMeshGroup.ts:12]: No. https://www.w3schools.com/jsref/jsref_concat_array.asp

File: src/parsers/i3d/CustomFileReader.ts (hunk 28 above)
Comment (anderscognite) [inline src/parsers/i3d/CustomFileReader.ts]: Types should be in upper camel case.

File: src/parsers/i3d/sharedFileParserTypes.ts [hunk 35]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2019 Cognite AS
 
 import FibonacciDecoder from '../FibonacciDecoder';
-import { BYTES_PER_NODE_ID, geometryNameType } from './parserParameters';
+import { BYTES_PER_NODE_ID, fileGeometryNameType } from './parserParameters';
```
Comment (anderscognite) [inline src/parsers/i3d/sharedFileParserTypes.ts]: Types should be in upper camel case.

File: src/parsers/protobuf/main.ts [hunk 36]
```diff
@@ -162,7 +161,7 @@ export default async function parseProtobuf(
 
   const rootSector = sectors['0/'];
   for (const sector of rootSector.traverseSectors()) {
-    mergeInstancedMeshes(sector, 2500, sceneStats, treeIndexNodeIdMap);
+    mergeInstancedMeshes(rootSector, sceneStats, treeIndexNodeIdMap);
```
Comment (anderscognite) [inline src/parsers/protobuf/main.ts]: This should be outside the loop.

File: src/parsers/protobuf/parseMergedMeshes.ts [hunk 37]
```diff
@@ -49,7 +49,8 @@ export default function parse(data: ParseData): MergedMeshGroup {
 
       const { triangleCount } = node;
       globalColor.setHex(color.rgb);
-      mergedMesh.mappings.add(triangleOffset, triangleCount, nodeId, treeIndex);
+      // diagonal size is calculated later
```
Comment (anderscognite) [inline src/parsers/protobuf/parseMergedMeshes.ts]:
```suggestion
      // size is calculated later
```

File: src/parsers/protobuf/parseInstancedMeshes.ts [hunk 38]
```diff
@@ -49,7 +49,8 @@ function createCollection(
     const color = property.color != null ? property.color : { rgb: 0xff00ff };
     globalColor.setHex(color.rgb);
     parseInstancedMeshTransformMatrix(globalMatrix, transformMatrix);
-    collection.addMapping(nodeId, treeIndex, globalMatrix);
+    // diagonal size is calculated later
```
Comment (anderscognite) [inline src/parsers/protobuf/parseInstancedMeshes.ts]:
```suggestion
    // size is calculated later
```

**Review by anderscognite — CHANGES_REQUESTED** [review]
ALMOST!!!!!!!!1111

**Review by anderscognite — APPROVED** [review]
